boards/litex_sim/src/main.rs

```diff
@@ -0,0 +1,420 @@
+use kernel::common::StaticRef;
+use kernel::component::Component;
+use kernel::hil::time::{Alarm, Timer};
+use kernel::Chip;
+use kernel::InterruptService;
+use kernel::Platform;
+use kernel::{create_capability, debug, static_init};
+use rv32i::csr;
+
+mod io;
+mod litex_generated_constants;
+
+// This module contains the LiteX SoC configuration options, register
+// positions, interrupt mappings and other implementation details of
+// the generated bitstream.
+//
+// Its values are used throughout the file, hence import it under a
+// short name.
+use litex_generated_constants as socc;
+
+/// Structure for dynamic interrupt mapping, depending on the SoC
+/// configuration
+///
+/// This struct is deliberately kept in the board crate. Because of
+/// the configurable nature of LiteX, it does not make sense to define
+/// a default interrupt mapping, as the interrupt numbers are
+/// generated sequentially for all softcores.
+struct LiteXSimInterruptablePeripherals {
+    uart0: &'static litex_vexriscv::uart::LiteXUart<'static, socc::SoCRegisterFmt>,
+    timer0: &'static litex_vexriscv::timer::LiteXTimer<
+        'static,
+        socc::SoCRegisterFmt,
+        socc::ClockFrequency,
+    >,
+    ethmac0: &'static litex_vexriscv::liteeth::LiteEth<'static, socc::SoCRegisterFmt>,
+}
+
+impl InterruptService<()> for LiteXSimInterruptablePeripherals {
+    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
+        match interrupt as usize {
+            socc::UART_INTERRUPT => {
+                self.uart0.service_interrupt();
+                true
+            }
+            socc::TIMER0_INTERRUPT => {
+                self.timer0.service_interrupt();
+                true
+            }
+            socc::ETHMAC_INTERRUPT => {
+                self.ethmac0.service_interrupt();
+                true
+            }
+            _ => false,
+        }
+    }
+
+    unsafe fn service_deferred_call(&self, _task: ()) -> bool {
+        false
+    }
+}
+
+const NUM_PROCS: usize = 4;
+
+// Actual memory for holding the active process structures. Need an
+// empty list at least.
+static mut PROCESSES: [Option<&'static dyn kernel::procs::ProcessType>; NUM_PROCS] =
+    [None; NUM_PROCS];
+
+// Reference to the chip and UART hardware for panic dumps
+struct LiteXSimPanicReferences {
+    chip: Option<
+        &'static litex_vexriscv::chip::LiteXVexRiscv<
+            VirtualMuxAlarm<
+                'static,
+                litex_vexriscv::timer::LiteXAlarm<
+                    'static,
+                    'static,
+                    socc::SoCRegisterFmt,
+                    socc::ClockFrequency,
+                >,
+            >,
+            LiteXSimInterruptablePeripherals,
+        >,
+    >,
+    uart: Option<&'static litex_vexriscv::uart::LiteXUart<'static, socc::SoCRegisterFmt>>,
+}
+static mut PANIC_REFERENCES: LiteXSimPanicReferences = LiteXSimPanicReferences {
+    chip: None,
+    uart: None,
+};
```

Comment:
This seems like a slight improvement on my approach of re-initializing these items in io.rs. Unifies what we do for CHIP with everything that has to be used in panic context, and makes it clear what is happening. I may try to port this to the other boards at some point

Reply:
I'm not entirely sure whether it's a large improvement, but at one stage while porting I had 5 different `static`s which would need to be passed to the panic handler, where the benefits of this approach clearly showed. :smile: